Write daily IRs for incremental aggregation

kambstreat opened the pull request:
## Summary
<!-- Overview of the changes involved in the PR -->

This PR address the following CHIP : https://github.com/airbnb/chronon/issues/984


## Why / Goal
<!-- Use cases and qualitative impact / opportunities unlocked -->

The goal of the PR is to reduce the time taken to compute aggregations. Added daily level partial aggregates for batch feature computation, which avoids reading past day's event level data. 

With this change, we see, 4x improvement for computing features. (except first time run ) 
## Test Plan
<!-- What was the process for testing the PR. How would someone extending / refactoring the work know it works. Not all
of these apply to every PR. -->
- [x] Added Unit Tests
- [x] Covered by existing CI
- [x] Integration tested

## Checklist
- [ ] Documentation update

## Reviewers

kambstreat committed:
code to write daily irs

kambstreat committed:
store incremental agg and compute final IRs

kambstreat committed:
Store hops to inc tables

pengyu-hou commented on aggregator/src/main/scala/ai/chronon/aggregator/row/RowAggregator.scala:
```diff
@@ -70,6 +70,11 @@ class RowAggregator(val inputSchema: Seq[(String, DataType)], val aggregationPar
     .toArray
     .zip(columnAggregators.map(_.irType))
 
+  val incSchema = aggregationParts
```
I am thinking that we should use a full `incremental` in the code and we can keep `inc` as the suffix for the table so the table names are not getting too long. What do you think? 
```suggestion
  val incrementalSchema = aggregationParts
```

kambstreat replied:
yes. good to use full name.

kambstreat committed:
add code changes to generate final output from IR for AVG

kambstreat committed:
change function structure and variable names

kambstreat committed:
remove unused functions

kambstreat committed:
change function defs

kambstreat committed:
make changes

kambstreat committed:
change function order

kambstreat committed:
add new field is_incremental to python api

kambstreat committed:
get argument for isIncremental in scala spark backend

kambstreat committed:
add unit test for incremental groupby

pengyu-hou commented on spark/src/test/scala/ai/chronon/spark/test/GroupByTest.scala:
```diff
@@ -694,4 +696,96 @@ class GroupByTest {
+    implicit val tableUtils = TableUtils(spark)
+
+    val namespace = "test_incremental_group_by"
+
```
```suggestion
tableUtils.createDatabase(namespace)
```

pengyu-hou commented on api/src/main/scala/ai/chronon/api/Extensions.scala:
```diff
@@ -98,6 +98,7 @@ object Extensions {
     def cleanName: String = metaData.name.sanitize
 
     def outputTable = s"${metaData.outputNamespace}.${metaData.cleanName}"
+    def incrementalOutputTable = s"${metaData.outputNamespace}.${metaData.cleanName}_inc"
```
```suggestion
    def incrementalOutputTable = s"${metaData.outputNamespace}.${metaData.cleanName}_daily_inc"
```

kambstreat committed:
reuse table ccreation

kambstreat committed:
Update GroupByTest

kambstreat committed:
Add GroupByTest for events

kambstreat committed:
changes for incrementalg

kambstreat committed:
resolve merge conflicts

kambstreat committed:
add last hole logic for incrementnal bacckfill

kambstreat committed:
fix syntax

kambstreat committed:
fix bug : backfill only for missing holes

kambstreat committed:
fix none error for inc Table

kambstreat committed:
add incremental table queryable range

kambstreat committed:
add logging for tableUtils

kambstreat committed:
add log

kambstreat committed:
fill incremental holes

kambstreat committed:
modify incremental aggregation parts

kambstreat committed:
remove logs for debugging

abbywh commented on aggregator/src/main/scala/ai/chronon/aggregator/base/SimpleAggregators.scala:
```diff
@@ -117,6 +117,57 @@ class UniqueCount[T](inputType: DataType) extends SimpleAggregator[T, util.HashS
   }
 }
 
+class AverageIR extends SimpleAggregator[Array[Any], Array[Any], Double] {
```
It looks to me like this mostly exists to serialize aggregators. Is there something more generic we can do? @nikhilsimha and I talked about https://fory.apache.org/ for example, but it could be even simpler to moving this logic into a different class abstraction

pengyu-hou replied:
I agree that if we could somehow make it more general, then we would not have to implement for each operation.

pengyu-hou replied:
also cc @nikhil-zlai 😁

pengyu-hou replied:
I was thinking of sticking with Avro because it is what we used for the online serving path so that we can keep the ser//de logic same.

kambstreat committed:
fix output schema from incremenntal aggregations. Added unit tests

kambstreat committed:
resolve merge conflicts

kambstreat committed:
resolve merge conflict

kambstreat committed:
add test case to test struct of Average

kambstreat committed:
add option for isIncremental for backward compatibility

kambstreat committed:
resolve merge conflicts

kambstreat committed:
fix count operation from incremental IRS

kambstreat committed:
use saw tooth aggregator to compute from daily IRs

kambstreat committed:
remove Average IR

kambstreat committed:
remove empty spaces and unused functions

kambstreat committed:
testing subset of aggregations

kambstreat committed:
remove duplicate test

kambstreat committed:
add unit tests for all aggregations

kambstreat committed:
convert spark datatype to java

kambstreat committed:
resolve merge conflicts

kambstreat committed:
scala formatted

kambstreat committed:
fix for scala 2.13

kambstreat committed:
fix for scala 2.13

kambstreat committed:
change unit test

kambstreat committed:
add toSeq for scala 2.13 compatibility

kambstreat committed:
fix last/first tests

kambstreat committed:
add new test file for incremental aggregations

pengyu-hou committed:
fix failed ci

kambstreat committed:
remove obsolete comment

kambstreat committed:
Merge branch 'kchakka/incremental' of https://github.com/kambstreat/chronon into kchakka/incremental

kambstreat committed:
revert to master files tor spark_submit.sh and teams.json

kambstreat committed:
remove log statements during debug

kambstreat committed:
Add daily_inc suffix to incremental table

Co-authored-by: Pengyu Hou <3771747+pengyu-hou@users.noreply.github.com>
Signed-off-by: chaitanya <1847554+kambstreat@users.noreply.github.com>

kambstreat committed:
fix bug in flatten function

kambstreat committed:
Merge branch 'kchakka/incremental' of https://github.com/kambstreat/chronon into kchakka/incremental

kambstreat committed:
revert changes in GroupByTest

pengyu-hou committed:
Merge branch 'main' of github.com:airbnb/chronon into kchakka/incremental

kambstreat committed:
remove comment

pengyu-hou committed:
Merge branch 'main' into kchakka/incremental

pengyu-hou committed:
Merge branch 'kchakka/incremental' of https://github.com/kambstreat/chronon into kchakka/incremental

pengyu-hou committed:
Merge branch 'main' into kchakka/incremental

pengyu-hou committed:
Merge branch 'kchakka/incremental' of https://github.com/kambstreat/chronon into kchakka/incremental

pengyu-hou committed:
Strengthen incremental backfill test coverage and fix bugs in GroupBy.scala

  GroupBy.scala — bug fixes

  - Fix aggregationParts returning empty when no holes exist: Previously derived from the hole-filling loop, so when the incremental table was already up-to-date, convertIncrementalDfToHops received an empty list
  and produced wrong hops. Now derived directly from groupByConf unconditionally.
  - Fix maxWindow.get NPE: Replaced with getOrElse(throw ...) that gives a clear error message when incremental mode is used on a GroupBy with no windowed aggregations.
  - Remove unused imports: Operation and com.google.common.collect.Table.
  - Fix typo: "aggregatiosn" → "aggregations" in scaladoc.

  GroupByIncrementalTest.scala — test improvements

  - Enable UNIQUE_COUNT and BOUNDED_UNIQUE_COUNT: Uncommented both operations in testIncrementalBasicAggregations. Comparison converts array IRs to their sizes (since element ordering/MD5 hashing differs between
  Chronon and SQL) and validates against count(distinct ...) from SQL.
  - Rewrite testIncrementalStatisticalAggregations: Replaced schema-only checks with a full incremental-vs-non-incremental snapshotEvents comparison for SKEW, KURTOSIS, APPROX_PERCENTILE, APPROX_UNIQUE_COUNT, and
  APPROX_HISTOGRAM_K.
  - Remove redundant column-existence assertions: Dropped assertTrue("IR must contain X") checks that were superseded by the value comparison.
  - Remove unused BinaryType import.

pengyu-hou committed:
Merge branch 'main' into kchakka/incremental

pengyu-hou committed:
Merge branch 'main' into kchakka/incremental
